### ISSUE TITLE
Z3 very slow on simple floating-point VC

### ISSUE THREAD

**Issue by yannickmoy** [Z3Prover/z3#823]
We are trying to use Z3 to prove critical code involving floating-point arithmetic, and we've hit a few cases where Z3 is too slow to be usable, even on small VCs with "simple" (if this can be defined!) properties, like monotonicity of addition here. I'm attaching a manually crafted VC reproducing the problem, which a very recent snapshot of Z3 (Nov 21 2016) takes 8 minutes on a modern machine (Core i7 with 2,8 GHz and 16 GB RAM) to prove 'unsat'. We would very much appreciate any recommendation on how to use Z3 to prove such properties (use specific switches? add specific assertions?), and any explanation that helps us decide when a property will not be provable by Z3 in a reasonable time due to the algorithms used.

To understand the VC below:
We have that x1 and x2 are in the range -2,000,000 .. 2,000,000
and that y is between -1 and 1 (all in floats)
and that x1 is less than or equal to x2
and we would like to prove that (x1 + y) is less than or equal to (x2 + y)

The calls to function 'finite' are probably useless, but were added in the hope of helping Z3.

CVC4 proves this property immediately (with a different encoding as CVC4 does not support natively floating-point numbers).

(set-logic QF_FPBV)
(set-info :source |SPARK inspired floating point problems|)
(set-info :smt-lib-version 2.5)
(set-info :category crafted)
(set-info :status unsat)

(define-const low_bound  Float32 ((_ to_fp 8 24) RNE (- 2000000.0)))
(define-const high_bound Float32 ((_ to_fp 8 24) RNE 2000000.0))

(define-const minus_one  Float32 ((_ to_fp 8 24) RNE (- 1.0)))
(define-const one        Float32 ((_ to_fp 8 24) RNE 1.0))

(define-fun finite ((x Float32)) Bool
  (or (fp.isZero x) (fp.isNormal x) (fp.isSubnormal x)))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Parameters
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(declare-const x1 Float32)
(assert (fp.geq x1 low_bound))
(assert (fp.leq x1 high_bound))
(assert (finite x1))

(declare-const x2 Float32)
(assert (fp.geq x2 low_bound))
(assert (fp.leq x2 high_bound))
(assert (finite x2))

(declare-const y Float32)
(assert (fp.geq y minus_one))
(assert (fp.leq y one))
(assert (finite y))

(assert (fp.leq x1 x2))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Implementation
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(define-const new_x1 Float32 (fp.add RNE x1 y))
(define-const new_x2 Float32 (fp.add RNE x2 y))

(assert (finite new_x1))
(assert (finite new_x2))

(assert (not (fp.leq new_x1 new_x2)))
(check-sat)
(exit)

**Comment by wintersteiger** [comment]
I'm afraid this is one of those cases that looks simple but isn't easy for the solver. Z3 translates this into bit-vectors, then bit-blasts, then runs a SAT solver and almost all the time is spent in the SAT solver. Implicitly, it enumerates lots of different cases before it finishes. That strategy is good for finding counter-examples, but it will always be inefficient for finding proofs of unsatisfiability over large bit-width floats.

We have techniques to handle small cores in big problems and for problems that use large floating-point numbers when they need only small ones, but neither of them applies here. In general, proving non-trivial things about complicated sets of floating-point numbers, will always be slow, except if the bit-width is very small, or we provide the right lemmas to Z3 so that it can cut some proofs short. 

For this particular example, it's fairly easy to check that the solver essentially enumerates. For instance, if we replace one/minus_one like so:

    (define-const one Float32 (fp #b0 #x00 #b00000000000000000000001))
    (define-const minus_one Float32 (fp #b1 #x00 #b00000000000000000000001))

(i.e., to the smallest non-zero float), then the solver takes only a couple seconds. And, as we move the '1' bit further to the left, it takes more and more time. 

I think that to a human this problem looks 'easy' because we implicitly proof it by induction (e.g., +1 means it will also hold for +2), but a SAT solver isn't able to do that. In general we might require quantified lemmas/axioms to do that efficiently, but that would take us into a whole different set of challenges.

**Comment by yannickmoy** [comment]
Thanks Christoph, that's very helpful to know what we can expect from Z3. To prove this kind of VCs, it is thus necessary to reason with axioms on floats, which may either be provided to the prover or whose particular instance on arguments may be applied manually. We will pursue these solutions then.

**Comment by wintersteiger** [comment]
Closing this issue as resolved, feel free to open a new one if new problems with floats come up!

**Comment by OmerSakar** [comment]
Dear @yannickmoy,

I had a question regarding to your statement (in the original question): 

> CVC4 proves this property immediately (with a different encoding as CVC4 does not support natively floating-point numbers).

I am currently experimenting with floats in Z3 and I was interested in this comment of yours. You mention a different encoding. I was wondering what encoding this was. Do you have this problem encoded in the different encoding?

Given that it has been 5 years, I would guess you do not have it. However, if you can point me in the right direction, that would also help a lot.

**Comment by yannickmoy** [comment]
@OmerSakar The encoding we were using for CVC4 at the time was using indirectly real numbers to represent floats. So you have an abstract type for floats, with a projection from floats to real numbers, so that it is correct for +0/-0 (which are different floats but project to the same real number 0.0). We do not have to worry in SPARK about +Inf/-Inf/NaN as we reject such values already. This is essentially the encoding we're still using for prover Alt-Ergo. For CVC4, we switches to the native encoding for floats after it was added in CVC4, so it has the same issues as Z3 now. 

We're using Why3 platform to driver provers and chose the translation strategy most appropriate for each prover, in case you are interested in doing the same. Out of curiosity, what is the context of your question?

**Comment by OmerSakar** [comment]
Thank you for the answer:) 

The context of my question is that we want to have float support in [VerCors](https://vercors.ewi.utwente.nl/). VerCors uses [Viper](https://www.pm.inf.ethz.ch/research/viper.html) which uses Z3. Viper recently added support for bit-vectors and floats. This leads me to Z3's floats.

When I was playing around with simple examples, the verification times were very high. I tried this example and it still takes a long time (for the same reason). So I started investigating why this was happening, which led me here.
